diff --git a/src/gdb_main.c b/src/gdb_main.c
index 1f50f05..4167222 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -117,7 +117,10 @@ int gdb_main_loop(struct target_controller *tc, bool in_syscall)
 	while (1) {
 		SET_IDLE_STATE(1);
 		size_t size = gdb_getpacket(pbuf, BUF_SIZE);
-		SET_IDLE_STATE(0);
+		// If port closed and target detached, stay idle
+		if (!(pbuf[0] == 0x04) || cur_target) {
+			SET_IDLE_STATE(0);
+		}
 		switch (pbuf[0]) {
 		/* Implementation of these is mandatory! */
 		case 'g': { /* 'g': Read general registers */
diff --git a/src/platforms/blackpillv2/platform.c b/src/platforms/blackpillv2/platform.c
index 207ad93..6493c22 100644
--- a/src/platforms/blackpillv2/platform.c
+++ b/src/platforms/blackpillv2/platform.c
@@ -115,8 +115,22 @@ void platform_init(void)
 	OTG_FS_GCCFG &= ~(OTG_GCCFG_VBUSBSEN | OTG_GCCFG_VBUSASEN);
 }
 
-void platform_nrst_set_val(bool assert) { (void)assert; }
-bool platform_nrst_get_val(void) { return false; }
+void platform_nrst_set_val(bool assert)
+{
+	if (assert) {
+		gpio_mode_setup(NRST_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, NRST_PIN);
+		gpio_set_output_options(NRST_PORT, GPIO_OTYPE_OD, GPIO_OSPEED_2MHZ, NRST_PIN);
+		gpio_clear(NRST_PORT, NRST_PIN);
+	} else {
+		gpio_mode_setup(NRST_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, NRST_PIN);
+		gpio_set(NRST_PORT, NRST_PIN);
+	}
+}
+
+bool platform_nrst_get_val(void)
+{
+	return gpio_get(NRST_PORT, NRST_PIN) == 0;
+}
 
 const char *platform_target_voltage(void)
 {
diff --git a/src/platforms/blackpillv2/platform.h b/src/platforms/blackpillv2/platform.h
index 11dac32..92dd608 100644
--- a/src/platforms/blackpillv2/platform.h
+++ b/src/platforms/blackpillv2/platform.h
@@ -75,9 +75,9 @@
 #define LED_PORT GPIOC
 #define LED_PORT_UART GPIOA
 #define LED_UART GPIO1
-#define LED_IDLE_RUN GPIO15
+#define LED_IDLE_RUN GPIO13
 #define LED_ERROR GPIO14
-#define LED_BOOTLOADER GPIO13
+#define LED_BOOTLOADER GPIO15
 
 #define USBUSART USART1
 #define USBUSART_CR1 USART1_CR1
diff --git a/src/platforms/stm32/blackpillv2.ld b/src/platforms/stm32/blackpillv2.ld
index 4dbc774..e2dc947 100644
--- a/src/platforms/stm32/blackpillv2.ld
+++ b/src/platforms/stm32/blackpillv2.ld
@@ -17,11 +17,13 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+/* least common denominator of STM32F401CCU6, STM32F401CEU6 and STM32F411CEU6 */
+
 /* Define memory regions. */
 MEMORY
 {
-	rom (rx) : ORIGIN = 0x08000000, LENGTH = 512K
-	ram (rwx) : ORIGIN = 0x20000000, LENGTH = 96K
+	rom (rx) : ORIGIN = 0x08000000, LENGTH = 256K
+	ram (rwx) : ORIGIN = 0x20000000, LENGTH = 64K
 }
 
 /* Include the common ld script from libopenstm32. */
