diff --git a/src/platforms/f4discovery/platform.h b/src/platforms/f4discovery/platform.h
index dbec221..d034fbc 100644
--- a/src/platforms/f4discovery/platform.h
+++ b/src/platforms/f4discovery/platform.h
@@ -76,9 +76,9 @@
 #define LED_PORT GPIOC
 #define LED_PORT_UART GPIOA
 #define LED_UART GPIO1
-#define LED_IDLE_RUN GPIO15
+#define LED_IDLE_RUN GPIO13
 #define LED_ERROR GPIO14
-#define LED_BOOTLOADER GPIO13
+#define LED_BOOTLOADER GPIO15
 
 #define USBUSART USART1
 #define USBUSART_CR1 USART1_CR1
