diff --git a/src/platforms/f4discovery/platform.c b/src/platforms/f4discovery/platform.c
index 4233f10..d25698d 100644
--- a/src/platforms/f4discovery/platform.c
+++ b/src/platforms/f4discovery/platform.c
@@ -131,8 +131,22 @@ void platform_init(void)
 #endif
 }
 
-void platform_srst_set_val(bool assert) { (void)assert; }
-bool platform_srst_get_val(void) { return false; }
+void platform_srst_set_val(bool assert)
+{
+	if (assert) {
+		gpio_mode_setup(SRST_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, SRST_PIN);
+		gpio_set_output_options(SRST_PORT, GPIO_OTYPE_OD, GPIO_OSPEED_2MHZ, SRST_PIN);
+		gpio_clear(SRST_PORT, SRST_PIN);
+	} else {
+		gpio_mode_setup(SRST_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, SRST_PIN);
+		gpio_set(SRST_PORT, SRST_PIN);
+	}
+}
+
+bool platform_srst_get_val()
+{
+	return gpio_get(SRST_PORT, SRST_PIN) == 0;
+}
 
 const char *platform_target_voltage(void)
 {
@@ -157,4 +171,4 @@ void platform_target_set_power(bool power)
 {
 	gpio_set_val(PWR_BR_PORT, PWR_BR_PIN, !power);
 }
-#endif
\ No newline at end of file
+#endif
